Sources/SwiftProtobuf/AnyMessageStorage.swift

```diff
@@ -301,163 +323,173 @@ extension AnyMessageStorage {
-
-    // If both have messages, check if they are the same.
-    if case .message(let myMsg) = state, case .message(let otherMsg) = other.state, type(of: myMsg) == type(of: otherMsg) {
-      // Since the messages are known to be same type, we can claim both equal and
-      // not equal based on the equality comparison.
-      return myMsg.isEqualTo(message: otherMsg)
-    }
-
-    // If both have serialized data, and they exactly match; the messages are equal.
-    // Because there could be map in the message, the fact that the data isn't the
-    // same doesn't always mean the messages aren't equal. Likewise, the binary could
-    // have been created by a library that doesn't order the fields, or the binary was
-    // created using the appending ability in of the binary format.
-    if case .binary(let myValue) = state, case .binary(let otherValue) = other.state, myValue == otherValue {
-      return true
-    }
-
-    // If both have contentJSON, and they exactly match; the messages are equal.
-    // Because there could be map in the message (or the JSON could just be in a different
-    // order), the fact that the JSON isn't the same doesn't always mean the messages
-    // aren't equal.
-    if case .contentJSON(let myJSON, _) = state,
-       case .contentJSON(let otherJSON, _) = other.state,
-       myJSON == otherJSON {
-      return true
-    }
-
-    // Out of options. To do more compares, the states conversions would have to be
-    // done to do comparisions; and since equality can be used somewhat removed from
-    // a developer (if they put protos in a Set, use them as keys to a Dictionary, etc),
-    // the conversion cost might be to high for those uses.  Give up and say they aren't equal.
-    return false
-  }
+      #if swift(>=4.2)
+        // Can't use _valueData for a few reasons:
+        // 1. Since decode is done on demand, two objects could be equal
+        //    but created differently (one from JSON, one for Message, etc.),
+        //    and the hash values have to be equal even if we don't have data
+        //    yet.
+        // 2. map<> serialization order is undefined. At the time of writing
+        //    the Swift, Objective-C, and Go runtimes all tend to have random
+        //    orders, so the messages could be identical, but in binary form
+        //    they could differ.
+        public func hash(into hasher: inout Hasher) {
+          if !_typeURL.isEmpty {
+            hasher.combine(_typeURL)
+          }
+        }
+      #else  // swift(>=4.2)
+        var hashValue: Int {
+          var hash: Int = i_2166136261
+          if !_typeURL.isEmpty {
+            hash = (hash &* i_16777619) ^ _typeURL.hashValue
+          }
+          return hash
+        }
+      #endif
+
+      func isEqualTo(other: AnyMessageStorage) -> Bool {
+            if _typeURL != other._typeURL { return false }
+
+            // Since the library does lazy Any decode, equality is a very hard problem.
```

Comment:
Don't these comments go past 80 columns?  It seems to me that word wrapping the paragraphs is not unreasonable, but we'd have to be smart about not disturbing ASCII art, etc.  I have some ideas about that if anyone's interested.

Reply:
They do—non-documentation comments that exceed the column limit are something we don't have a great way to handle yet, so I'm definitely interested in your thoughts on that.

Separately, for documentation comments, one thing I've considered is parsing the Markdown and then rerendering it back out with the desired column limit (which is why the whole `CommonMark` module, wrapping `cmark`, exists, even though I haven't gotten far enough to use it yet). One drawback is that cmark doesn't preserve the original characters used for emphasis, strong text, bulleted lists, etc., so if the user used underscores to emphasize a word they would get converted to asterisks as a result of formatting it. Maybe we'd be OK with that kind of canonicalization? But it also feels a bit heavy-handed.